pkg/scheduler/ugm/manager_test.go

```diff
@@ -702,6 +702,93 @@ func TestDecreaseTrackedResourceForGroupTracker(t *testing.T) {
+	}
+	emptyUsage, err := resources.NewResourceFromConf(map[string]string{"memory": "0", "vcore": "0"})
+	if err != nil {
+		t.Errorf("new resource create returned error or wrong resource: error %t, res %v", err, emptyUsage)
+	}
+
+	// Assert inital ugm state
+	assertUgmTrackedResourceCount(t, 0, 2)
+
+	// Assert inital GroupTracker state
+	assetGroupTrackerResources(t, "group1", 0, nil)
+	assetGroupTrackerResources(t, "group2", 0, nil)
+
+	// Increase resources to different application, different queue, check success
+	increaseTrackedThenAssert(t, "root.parent.leaf1", TestApp1, usage, userGroup, true)
+	increaseTrackedThenAssert(t, "root.parent.leaf2", TestApp2, usage, userGroup, true)
+
+	// Assert ugm state after increase resources
+	assertUgmTrackedResourceCount(t, 1, 2)
+
+	// Assert UserTracker/GroupTracker, new UserTracker should be careated.
+	assetUserTrackerResources(t, user, 2, resources.Multiply(usage, 2))
+	assetUserTrackerAppGroupTrackerMapping(t, user, TestApp1, "group2")
+	assetUserTrackerAppGroupTrackerMapping(t, user, TestApp2, "group1")
+	assertGroupTrackerResources(t, "group1", TestApp2, usage, 1, "user")
+	assertGroupTrackerResources(t, "group2", TestApp1, usage, 1, "user")
+
+	assertGroupTrackerQueuePathResource(t, "group1", "root.parent.leaf1", nil)
+	assertGroupTrackerQueuePathResource(t, "group1", "root.parent.leaf2", usage)
+	assertGroupTrackerQueuePathResource(t, "group1", "root.parent", usage)
+	assertGroupTrackerQueuePathResource(t, "group1", "root", usage)
+	assertGroupTrackerQueuePathResource(t, "group2", "root.parent.leaf1", usage)
+	assertGroupTrackerQueuePathResource(t, "group2", "root.parent.leaf2", nil)
+	assertGroupTrackerQueuePathResource(t, "group2", "root.parent", usage)
+	assertGroupTrackerQueuePathResource(t, "group2", "root", usage)
+
+	// Test limit setting has been limiting new resources, check failed
+	increaseTrackedThenAssert(t, "root.parent.leaf1", TestApp1, usage, userGroup, false)
+	increaseTrackedThenAssert(t, "root.parent.leaf2", TestApp2, usage, userGroup, false)
+
+	// Remove apps from queue, check success
+	decreaseTrackedThenAssert(t, "root.parent.leaf1", TestApp1, usage, userGroup, true, true)
+	decreaseTrackedThenAssert(t, "root.parent.leaf2", TestApp2, usage, userGroup, true, true)
+
+	// Assert ugm/GroupTracker state after apps removed, UserTracker should be removed.
+	assertUgmTrackedResourceCount(t, 0, 2)
+	assertGroupTrackerResources(t, "group1", TestApp2, emptyUsage, 0, "")
+	assertGroupTrackerResources(t, "group2", TestApp1, emptyUsage, 0, "")
+
+	// root.parent.leaf1 TestApp1 -> was limited by group 2, root.parent.leaf1
+	// root.parent.leaf2 TestApp2 -> was limited by group 1, root.parent
+	assertGroupTrackerQueuePathResource(t, "group1", "root.parent.leaf1", nil)
+	// assertGroupTrackerQueuePathResource(t, "group1", "root.parent.leaf2", emptyUsage)  // -> Failed
+	// assertGroupTrackerQueuePathResource(t, "group1", "root.parent.leaf2", usage)       // -> Passed
```

Comment:
Hi @manirajv06,

I encountered a failed assert in Line 782 after decreasing resources, 
and I believe this is a bug related GroupTracker.DecreaseTrackedResource().

When I increased the resource in the 'root.parent.leaf2' queue,  (the limit was set on 'root.parent')
the resource was tracked in the following queueTracker:
- root.parent.leaf2
- root.parent
- root

However, when I decreased the resource from 'root.parent.leaf2',
it only removed resource in the following queueTracker:
- root.parent
- root

The root cause might be that there is no limit on 'root.parent.leaf2'.
The traversal of decreasing only decrease resources before root.parenet

Please correct me if I'm wrong.

Reply:
Sorry, please let me correct myself.  

- The behavior of GroupTracker.DecreaseTrackedResource() is correct
- The wrong state happened on QueueTracker.increaseTrackedResource() if the parent queue reach the limitation. for example: 

Before state:
- root.parent           (usage 10 memory, 10 vcore) + (limit 10 memory, 10 vcore)
- root.parent.leaf2 (usage 10 memory, 10 vcore) + (no limit)

After adding (10 memory, 10 vcore) resource to  root.parent.leaf2
- root.parent           (usage 10 memory, 10 vcore) 
- root.parent.leaf2.(usage 20 memory, 20 vcore)

Reply:
Maybe it's intentional because ugh doesn't care about those leaf queues without limits?

Reply:
Summary the issue:
Queue Setup:
- **root.parent with group1 limit**
- **root.parent.leaf2 with no limit**

Before assert:
- increaseTrackedResourceThenAssert(t, "root.parent.leaf2", TestApp2, usage, userGroup, true) //**success**
- increaseTrackedResourceThenAssert(t, "root.parent.leaf2", TestApp2, usage, userGroup, false)  
 **// reach 'root.parent' limit, but it still increase root.parent.leaf2's usage.**
- decreaseTrackedResourceThenAssert(t, "root.parent.leaf2", TestApp2, usage, userGroup, true, true) //**success**

Assert:
- assertGroupTrackerQueuePathResource(t, "group1", "root.parent.leaf2", emptyUsage) 
  **// -> Failed.  (Should Pass)**
- assertGroupTrackerQueuePathResource(t, "group1", "root.parent.leaf2", usage)              
  **//-> Passed. (Should Failed)**